DuckDuckGo/AppDelegate.swift

```diff
@@ -166,48 +156,15 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         }
 
         var shouldResetBookmarksSyncTimestamp = false
-
-        bookmarksDatabase.loadStore { [weak self] context, error in
-            guard let context = context else {
-                if let error = error {
-                    Pixel.fire(pixel: .bookmarksCouldNotLoadDatabase,
-                               error: error)
-                } else {
-                    Pixel.fire(pixel: .bookmarksCouldNotLoadDatabase)
-                }
-
-                if shouldPresentInsufficientDiskSpaceAlertAndCrash {
-                    return
-                } else {
-                    Thread.sleep(forTimeInterval: 1)
-                    fatalError("Could not create Bookmarks database stack: \(error?.localizedDescription ?? "err")")
-                }
-            }
-            
-            let legacyStorage = LegacyBookmarksCoreDataStorage()
-            legacyStorage?.loadStoreAndCaches()
-            LegacyBookmarksStoreMigration.migrate(from: legacyStorage,
-                                                  to: context)
-            legacyStorage?.removeStore()
-
-            do {
-                BookmarkUtils.migrateToFormFactorSpecificFavorites(byCopyingExistingTo: .mobile, in: context)
-                if context.hasChanges {
-                    try context.save(onErrorFire: .bookmarksMigrationCouldNotPrepareMultipleFavoriteFolders)
-                    if let syncDataProviders = self?.syncDataProviders {
-                        syncDataProviders.bookmarksAdapter.shouldResetBookmarksSyncTimestamp = true
-                    } else {
-                        shouldResetBookmarksSyncTimestamp = true
-                    }
-                }
-            } catch {
-                Thread.sleep(forTimeInterval: 1)
-                fatalError("Could not prepare Bookmarks DB structure")
-            }
-
-            WidgetCenter.shared.reloadAllTimelines()
+        if BookmarksDatabaseSetup.loadStoreAndMigrate(bookmarksDatabase: bookmarksDatabase,
+                                                      crashOnLoadStoreError: !shouldPresentInsufficientDiskSpaceAlertAndCrash) {
+            // MARK: post-Bookmarks migration logic
+            assertBookmarksMigrationHappensOnlyOnce()
+            shouldResetBookmarksSyncTimestamp = true
```

Comment:
I noticed that "syncDataProviders" can only be nil here because this immediately after launch before a value has been assigned, so I just reduced this to the logic that would be called anyway.